lib/utils.js

```diff
@@ -4,60 +4,62 @@ import { exec } from 'teen_process';
 import { waitForCondition } from 'asyncbox';
 import { getVersion } from 'appium-xcode';
 import { getDevices } from 'node-simctl';
-import { fs, system } from 'appium-support';
+import { fs } from 'appium-support';
 import { Certificate } from './certificate';
 import path from 'path';
 import Simulator from './simulator-xcode-6';
 
 
-const OSASCRIPT_TIMEOUT = 10000;
+const DEFAULT_SIM_SHUTDOWN_TIMEOUT = 10000;
 
-async function sendKillAll (appName) {
+// pgrep/pkill exit codes:
+// 0       One or more processes were matched.
+// 1       No processes were matched.
+// 2       Invalid options were specified on the command line.
+// 3       An internal error occurred.
+
+async function forceKillAll (appName) {
   try {
-    await exec('killall', [appName]);
-  } catch (err) {
-    // if no process exists we'll get an error that we expect.
-    // otherwise throw it again
-    if (err.stderr.indexOf(`No matching processes belonging to you were found`) === -1) {
-      throw err;
+    await exec('pkill', ['-9', '-x', appName]);
+  } catch (e) {
+    if (e.code === 1) {
+      log.debug(`${appName} is not running. Continuing...`);
+      return;
     }
+    log.error(`pkill error ${e.code} while trying to forcefully kill ${appName}`);
+    throw e;
   }
 }
 
-async function killAllSimulators (timeout = OSASCRIPT_TIMEOUT) {
+async function killAllSimulators (timeout = DEFAULT_SIM_SHUTDOWN_TIMEOUT) {
   log.debug('Killing all iOS Simulators');
 
-  let appName;
-  let xcodeVersion = await getVersion(true);
-  if (xcodeVersion.major >= 7) {
-    appName = 'Simulator';
-  } else {
-    appName = 'iOS Simulator';
-  }
+  const xcodeVersion = await getVersion(true);
+  const appName = xcodeVersion.major >= 7 ? 'Simulator' : 'iOS Simulator';
 
   try {
-    await exec('osascript', ['-e', `quit app "${appName}"`], {timeout});
-
-    // in Sierra the killing with AppleScript often pretends to work but fails
-    let macVersion = await system.macOsxVersion();
-    if (macVersion === '10.12') {
-      await sendKillAll(appName);
+    await exec('pgrep', ['-x', appName]);
+  } catch (e) {
+    if (e.code === 1) {
+      log.debug(`${appName} is not running. Continuing...`);
+      return;
     }
+    log.error(`pgrep error ${e.code} while detecting whether ${appName} is running. Trying to kill anyway...`);
+  }
+
+  try {
```

Comment:
this block of code is basically the same as what's factored out in a method above, maybe we can consolidate these into one method which also has a `-9`-type parameter?